docs/proposals/local-cluster.md

```diff
@@ -0,0 +1,486 @@
+proper kubectl binary and a symlink to it, and `admin.conf`). _Why
+this is needed:_ avoid downloading of uncompressed `kubectl` binary if
+the user's machine has slow/limited network connection; reduce the
+amount of logic in the shell script so it doesn't have to be updated
+often; provide an ability to pull extra binaries like `helm` in
+future. Not downloading any binaries isn't an option either because
+e.g. `kubectl` from k8s 1.5 can't properly handle deployments in k8s
+1.6.
+
+When `BUILD_*` options are enabled, `dind-cluster` startup scripts
+invokes appropriate `build/run.sh make ...` commands in Kubernetes
+source tree, preventing the build container from copying back the
+binaries using `KUBE_RUN_COPY_OUTPUT=n`. The control container then
+pulls the binaries directly from build container. _NOTE:_ extra tricks
+may be necessary in case `LOCAL_GO` is enabled.
+
+### Networking
+
+`local-cluster` utilizes named docker network
+`k8s-local-cluster`. This network is always created or replaced upon
+the cluster startup using address range specified via `DIND_SUBNET`
+configuration variable. `local-cluster` makes docker's internal DNS
+server available to all the nodes in the cluster, so nodes can be
+referred by their hostnames: `kube-master`, `kube-node-1`,
+`kube-node-2` and so on.
+
+### Cluster snapshot mechanism
+
+In order to reduce the amount of time needed to restart the cluster,
+`local-cluster` takes the snapshot of the cluster after setting it up.
+The snapshot is made by means of preserving each node's
+`/var/lib/docker` directory as a docker volume and saving each node
+container's filesystem in an image. The volumes and images created by
+snapshotting mechanism are removed during `./local-cluster.sh clean`.
+
+### Future directions
+
+The following features may be added later:
+
+* automatic PV creation like it's done in `minikube`, so most Helm
+  charts from
+  [kubernetes/charts](https://github.com/kubernetes/charts) repository
+  work out of the box
+* enabling feature gates for Kubernetes components
+* automatic installation of `helm` and `tiller`
+* support using `delve` to debug k8s components.  Should be an option
+  for building k8s with `GOGCFLAGS="-N -l"` plus `delve` binary in the
+  images. Possible problem: as of Go 1.7, trying to build e.g.
+  `kubectl` with `GOGCFLAGS="-N -l"` may result in stack overflow
+  in Go compiler.
+
+### CI
+
+A test matrix needs to be set up so each combination of supported
+Kubernetes version and supported CNI plugin is tested. In addition to
+this, there should be a test that builds Kubernetes binaries based on
+`master` branch of
+[kubernetes/kubernetes](https://github.com/kubernetes/kubernetes)
+repository. k8s `master` test should be executed daily so possible
+issues that break DIND can be tracked down in timely manner and with
+minimal use of `git bisect`.
+
+### Limitations
+
+There are problems with `btrfs` support because of a
```

Comment:
Well, `kubeadm-dind-cluster` seems also to work ok when host docker has `devicemapper` storage.